File: src/main/kotlin/io/github/ephemient/aoc2019/Day1.kt
package io.github.ephemient.aoc2019

class Day1(private val lines: List<String>) {
    private fun fuel(weight: Int): Int? = (weight / 3 - 2).takeIf { it > 0 }

    private fun fuels(weight: Int): Int = generateSequence(fuel(weight), ::fuel).sum()

    fun part1(): Int = lines.sumBy { fuel(it.toInt()) ?: 0 }

    fun part2(): Int = lines.sumBy { fuels(it.toInt()) }
}

File: src/aoc2019/day1.py
import fileinput


def fuel(weight):
    return weight // 3 - 2


def fuels(weight):
    while (weight := fuel(weight)) > 0:
        yield weight


def part1(lines):
    '''
    >>> part1(['12'])
    2
    >>> part1(['14'])
    2
    >>> part1(['1969'])
    654
    >>> part1(['100756'])
    33583
    '''
    return sum(fuel(int(line)) for line in lines)


def part2(lines):
    '''
    >>> part2(['14'])
    2
    >>> part2(['1969'])
    966
    >>> part2(['100756'])
    50346
    '''
    return sum(weight for line in lines for weight in fuels(int(line)))


parts = (part1, part2)

if __name__ == '__main__':
    lines = list(fileinput.input())
    print(part1(lines))
    print(part2(lines))
